File: samples/go/pricing.go
package bookstore

import "errors"

// Band is a quantity threshold and the discount it unlocks.
type Band struct {
	MinQuantity int
	Discount    float64
}

// BulkThresholds are the quantity discount bands, cheapest band first.
var BulkThresholds = []Band{
	{MinQuantity: 5, Discount: 0.05},
	{MinQuantity: 10, Discount: 0.10},
	{MinQuantity: 25, Discount: 0.18},
}

// TierMultipliers scales the final price by loyalty tier. Adding a Tier
// without adding an entry here silently prices at zero.
var TierMultipliers = map[Tier]float64{
	TierBronze: 1.00,
	TierSilver: 0.97,
	TierGold:   0.94,
}

// ErrEmptyCart is returned when there is nothing to price.
var ErrEmptyCart = errors.New("cannot price an empty cart")

// Validate rejects carts that cannot be priced.
func Validate(cart Cart) error {
	if len(cart.Lines) == 0 {
		return ErrEmptyCart
	}
	for _, line := range cart.Lines {
		if line.Quantity < 1 {
			return errors.New("quantities must be positive")
		}
	}
	return nil
}

// PricingService turns a cart into a payable amount.
type PricingService struct{}

// BasePrice sums unit price times quantity across all lines.
func (p PricingService) BasePrice(cart Cart) Money {
	total := NewMoney(0)
	for _, line := range cart.Lines {
		scaled := line.Book.UnitPrice.Scaled(float64(line.Quantity))
		sum, err := total.Add(scaled)
		if err != nil {
			return total
		}
		total = sum
	}
	return total
}

// ApplyBulkDiscount applies the quantity band, then the tier multiplier.
func (p PricingService) ApplyBulkDiscount(cart Cart, tier Tier) (Money, error) {
	if err := Validate(cart); err != nil {
		return Money{}, err
	}
	subtotal := p.BasePrice(cart)
	quantity := cart.TotalQuantity()

	discount := 0.0
	for _, band := range BulkThresholds {
		if quantity >= band.MinQuantity {
			discount = band.Discount
		}
	}
	return subtotal.Scaled((1.0 - discount) * TierMultipliers[tier]), nil
}

// ApplyCoupon takes 10% off for recognised codes.
func (p PricingService) ApplyCoupon(amount Money, code string) Money {
	if len(code) >= 4 && code[:4] == "BOOK" {
		return amount.Scaled(0.9)
	}
	return amount
}

File: samples/rust/src/pricing.rs
//! Pricing rules: bulk discounts and loyalty tier multipliers.

use crate::models::{Cart, Tier};
use crate::money::Money;

/// A quantity threshold and the discount it unlocks.
pub struct Band {
    pub min_quantity: i64,
    pub discount: f64,
}

/// Quantity discount bands, smallest threshold first.
pub const BULK_THRESHOLDS: [Band; 3] = [
    Band { min_quantity: 5, discount: 0.05 },
    Band { min_quantity: 10, discount: 0.10 },
    Band { min_quantity: 25, discount: 0.18 },
];

/// Scales the final price by loyalty tier. A new Tier needs an arm here.
pub fn tier_multiplier(tier: Tier) -> f64 {
    match tier {
        Tier::Bronze => 1.00,
        Tier::Silver => 0.97,
        Tier::Gold => 0.94,
    }
}

/// Rejects carts that cannot be priced.
pub fn validate(cart: &Cart) -> Result<(), String> {
    if cart.lines.is_empty() {
        return Err("cannot price an empty cart".to_string());
    }
    if cart.lines.iter().any(|line| line.quantity < 1) {
        return Err("quantities must be positive".to_string());
    }
    Ok(())
}

/// Turns a cart into a payable amount.
pub struct PricingService;

impl PricingService {
    /// Sums unit price times quantity across all lines.
    pub fn base_price(&self, cart: &Cart) -> Money {
        let mut total = Money::new(0);
        for line in &cart.lines {
            total = total.add(&line.book.unit_price.scaled(line.quantity as f64));
        }
        total
    }

    /// Applies the quantity band, then the tier multiplier.
    pub fn apply_bulk_discount(&self, cart: &Cart, tier: Tier) -> Result<Money, String> {
        validate(cart)?;
        let subtotal = self.base_price(cart);
        let quantity = cart.total_quantity();

        let mut discount = 0.0;
        for band in BULK_THRESHOLDS.iter() {
            if quantity >= band.min_quantity {
                discount = band.discount;
            }
        }
        Ok(subtotal.scaled((1.0 - discount) * tier_multiplier(tier)))
    }

    /// Takes 10% off for recognised coupon codes.
    pub fn apply_coupon(&self, amount: Money, code: &str) -> Money {
        if code.starts_with("BOOK") { amount.scaled(0.9) } else { amount }
    }
}
